File: SPOJ/TRVCOST - Travelling cost.cpp

#include <bits/stdc++.h>
#define MAXN 505
#define INF 9999999
using namespace std;
typedef long long int lli;

int m[MAXN+1][MAXN+1];

int N, U, Q;

void init(){
    int i, j;
    for( i = 0; i < MAXN; ++i ){
        for( j = 0; j < MAXN; ++j ){
            if( i != j ){
                m[i][j] = INF;
            } else {
                m[i][j] = 0;
            }
        }
    }
}

int main() {
    ios_base::sync_with_stdio(0); cin.tie(0);
    int i, j, k; int x, y, w; int maxN = 1; int V;
    int ans;
    cin >> N;
    init();
    for( i = 0; i < N; ++i ){
        cin >> x >> y >> w;
        maxN = max( maxN, max( x, y ) );
        m[x][y] = min( m[x][y], w );
        m[y][x] = min( m[y][x], w );
    }
    for( k = 0; k <= maxN; ++k ){
        for( i = 0; i <= maxN; ++i ){
            for( j = 0; j <= maxN; ++j ){
                m[i][j] = min( m[i][j] , m[i][k] + m[k][j] );
            }
        }
    }
    cin >> U;
    cin >> Q;
    for( i = 0; i < Q; ++i ){
        cin >> V;
        ans = min( m[U][V], m[V][U] );
        if( ans == INF ){
            cout << "NO PATH\n";
        } else {
            cout << ans << "\n";
        }
    }
    return 0;
}

File: SPOJ/TRVCOST - Travelling cost.rb
@MAXN = 505
@INF = 9999999

n = gets.to_i

m = Array.new(@MAXN+1) { Array.new(@MAXN+1) }

for i in 0..@MAXN
	for j in 0..@MAXN
		if i != j
			m[i][j] = @INF
		else 
			m[i][j] = 0
		end
	end
end

arr = Array.new(n){ Array.new() }

maxVal = 1

for i in 0..n-1
	s = gets
	s.split(" ").each { |e|
		arr[i] << e.to_i
	}
	a = arr[i][0]
	b = arr[i][1]
	c = arr[i][2]
	maxVal = [ maxVal , a , b ].max
	m[a][b] = [ m[a][b] , c ].min
	m[b][a] = [ m[b][a] , c ].min
end

for k in 0..maxVal
	for i in 0..maxVal
		for j in 0..maxVal
			m[i][j] = [ m[i][j] , m[i][k] + m[k][j] ].min
		end
	end
end

u = gets.to_i
q = gets.to_i

for i in 0..q-1
	v = gets.to_i
	ans = [ m[u][v] , m[v][u] ].min
	if ans == @INF 
		puts "NO PATH"
	else
		puts ans
	end
end
